File: Java/Leetcode/DP2D/UniquePaths.java
package com.brianprojects.Leetcode.DP2D;

public class UniquePaths {
    //The runtime complexity is: O(m*n) of O(n^2) which run linearly
    //The space complexity is O(m)
    public static int uniquePaths(int m, int n){
        int[][] path = new int[m+1][n+1];
        path[m-1][n-1] = 1;
        for(var r = m-1; r>=0; r--){
            for(var c =  n-1; c>=0;c--){
                path[r][c] = path[r][c] +path[r+1][c]+path[r][c+1];
            }
        }
        return path[0][0];
    }
}

File: Python/Leetcode_Python/2D-DP/UniquePaths.py
# Time Complexity: O(mxn) || Space Complexity: O(mxn)
def uniquePaths(m: int, n: int) -> int:
    if not m or not n:
        return -1

    paths = [[0 for _ in range(n + 1)] for _ in range(m + 1)]

    paths[1][1] = 1

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            paths[i][j] += paths[i - 1][j] + paths[i][j - 1]
    print(paths)
    return paths[m][n]


print(uniquePaths(3, 7))
